### ISSUE TITLE
Improve UI to add variable data to emails

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#7743]
**Describe the problem you would like to solve**
As part of #5717, many more variables can be used with email templates. However, the existing UI for adding variable data does not work well with this. It's hard to find the data you want in a long list. And when data is very long, like the abstract, it's hard to see what is being added.

![email-template-variables](https://user-images.githubusercontent.com/2306629/157068303-242bebb6-3899-4f3f-ae5a-4db007f1d62c.png)

**Describe the solution you'd like**
Redesign the UI to make it easier to find and insert variable data in an email. I imagine this would be a modal with a search/find interface that helped the user quickly find and insert the information they want.

**Who is asking for this feature?**
This is a UX improvement due to changes made in #5717.

**Comment by NateWr** [comment]
Short video of the new UI:

https://user-images.githubusercontent.com/2306629/159949762-b0ae0ab5-ba96-42b9-99f5-e20ac18f560b.mp4

The following PRs apply this change to the email signature and record decision fields. It replaces everywhere that the `preparedContent` property was used with the Vue.js editor field. It does not effect uses of the editor in old forms.

Find and test it in the UI library:
- `cd lib/ui-library && npm run serve`
- See the Form > FieldPreparedContent, Composer, and DecisionPage previews.

Find and test in OJS:
- Settings > Workflow > Emails > Email Signature
- Submission -> record any decision with an email notification

@ewhanson do you have time to do a code review on this?

PRs:
https://github.com/pkp/ui-library/pull/192
https://github.com/pkp/pkp-lib/pull/7791

Tests:
https://github.com/pkp/ojs/pull/3347
https://github.com/pkp/omp/pull/1090
https://github.com/pkp/ops/pull/255

**Comment by NateWr** [comment]
The variable substitution "pass" can happen again on the server. So in such cases, instead of passing the actual value to the UI component, we would pass the "variable name" as the value, eg:

```
{
  key: 'reviewUrl',
  value: '{$reviewUrl}',
}
```

When the email is submitted, the text will still be run through the variable substitution, so that `{$reviewUrl}` can be transformed into the `https://...` on the server. It should be fairly easy to rig things up this way when we have special cases to manage like this.

**Comment by ewhanson** [comment]
> @ewhanson do you have time to do a code review on this?

Yes, I'll have a look.

**Comment by asmecher** [comment]
Thanks, @NateWr -- in this case we're counting on the URL already being invisible to the user (unless they edit the link), right? i.e. for a "visible" variable, it would show up as {$someVariableName} in the email text and be replaced when sending?

**Comment by NateWr** [comment]
> for a "visible" variable, it would show up as {$someVariableName} in the email text and be replaced when sending?

:+1: exactly. A similar situation exists for the "notify reviewers" email. Since each reviewer is sent a separate email, the `{$recipientName}` needs to be left "raw" when the editor writes the email, and then replaced as each email is sent.

**Comment by asmecher** [comment]
This loses some flexibility we previously had in allowing late replacement of variables without forcing users to see sometimes-confusing `{$variableSlugs}`. (They often think these are glitches and try to fill them in manually.) See e.g. the "assign reviewer" message:

![image](https://user-images.githubusercontent.com/200411/159984853-5906d2ad-2dbf-481a-9981-0481ce823685.png)

In this case the "Response Due Date" variable can't be filled in with a specific date because the date picker is on the same form. So it was possible for the variable to appear in the message as a labeled widget, and the replacement with an actual value to happen on sending.

Just an example use case -- there are others.

**Comment by NateWr** [comment]
The new approach allows us to do whatever we want, including the approach you describe. We define what the "value" of any variable is when it is displayed in the editor, so we could make it `<span data-variable="reviewDueDate">Review Due Date</span>` and then convert that back to `{$reviewDueDate}` before processing the email body and sending it.

However, user testing has indicated that the placeholder styles that I implemented were not reducing user confusion. They were just as confused by uneditable text like `RESPONSE DUE DATE` as they were by the code-like syntax of`{$responseDueDate}`.

We'll have to revisit how to do this, but in the meantime I think our user testing suggests that reverting to `{$variables}` won't be a big problem. When we do revisit this, we can either pursue a different UI design that makes this clearer/easier. Or if we need, it is not difficult to replace these variables right there in the browser, so that when the user changes a review due date, that will update in the email itself. 

This is the benefit of the component-based approach of Vue.js. We don't have to build every use-case into the base component. We can manipulate the data at the parent level (for example, the form controlling reviewer assignment), and pass that down to the editor component, which doesn't need to concern itself too much with what the data looks like.

**Comment by ewhanson** [comment]
Hey @NateWr, this looks really good! I've left a few specific comments in the code but also had a few, more general comments (included below).

In the UI Library documentation, any preview page that contains a TinyMCE editor has the template code example styling overridden by one of the TinyMCE stylesheets (`content.inline.min.css`). This causes the code examples for these to look like the image below.

![Screen Shot 2022-03-28 at 16 27 54](https://user-images.githubusercontent.com/20844845/160505307-7d03e88e-5fe8-4a05-9e01-7056ff845efb.png)

Regarding, the `InsertContent` component/class more generally: If some variable don't have a description, it's not clear in the Vue component which key/variable the displayed content (value) is meant to represent (see screenshot below). For example, a URL or a repeated value like a name appearing more than once. If there isn't a description provided (or a translation of a description), could it fall back to the prepared content key? Or is there another way to ensure some context is always provided? I'm not sure if it ultimately matters in terms of the rendered email sent, but I imagine some of this could be confusing when a user first opens the InsertContent modal.

![Screen Shot 2022-03-28 at 12 08 47](https://user-images.githubusercontent.com/20844845/160506219-6a697adb-00f7-4001-894f-9fbfd57cb993.png)

And finally, this is not directly related to the work in this PR, but I came across it while reviewing it: In the add reviewer email template (`REVIEW_REQUEST`), the email template uses `reviewAssignmentUrl` and the prepared content variable uses `submissionReviewUrl` as the key. This was leading to the reviewer request email including `{$reviewAssignmentUrl}` as part of the final rendered text as it wasn't getting replaced.

Thanks!

**Comment by NateWr** [comment]
Thanks @ewhanson! I think I've replied to each of your comments. The tests are running now. They failed before, but it looked like maybe a one-off. :crossed_fingers: 

> In the UI Library documentation, any preview page that contains a TinyMCE editor has the template code example styling overridden by one of the TinyMCE stylesheets

I know, it's really annoying! :laughing: It's on my general-try-to-fix-this-for-3.4-if-theres-time todo list.

> If some variable don't have a description, it's not clear in the Vue component which key/variable the displayed content (value) is meant to represent

Yeah, I'd like to solve this by ensuring that every variable has a description. There are a lot of moving parts right now with email variables, so I've decided to leave this until we do 3.4 testing to resolve. I think little things like this will continue to crop up as we go along and implement more email variables.

> In the add reviewer email template (REVIEW_REQUEST), the email template uses reviewAssignmentUrl and the prepared content variable uses submissionReviewUrl as the key.

Vitaliy is doing a lot of work right now on the review email templates, and that includes migrations to email variable names for consistency (#7592). I'm going to wait until that is in to check this stuff, but good to keep it in mind as something to test when we get to that stage.
